pilot/pkg/config/kube/gateway/deploymentcontroller.go

```diff
@@ -450,6 +399,58 @@ func (d *DeploymentController) configureIstioGateway(log *istiolog.Scope, gw gat
 	return nil
 }
 
+func (d *DeploymentController) setLabelOverrides(gw gateway.Gateway, input TemplateInput) {
+	// TODO: Codify this API (i.e how to know if a specific gateway is an Istio waypoint gateway)
+	isWaypointGateway := strings.Contains(string(gw.Spec.GatewayClassName), "waypoint")
+
+	var hasAmbientLabel bool
+	if _, ok := gw.Labels[constants.DataplaneModeLabel]; ok {
+		hasAmbientLabel = true
+	}
+	if _, ok := input.InfrastructureLabels[constants.DataplaneModeLabel]; ok {
+		hasAmbientLabel = true
+	}
+	// If no ambient redirection label is set explicitly, explicitly disable.
+	// TODO this sprays ambient annotations/labels all over EVER gateway resource (serviceaccts, services, etc)
+	if features.EnableAmbientWaypoints && !isWaypointGateway && !hasAmbientLabel {
+		input.InfrastructureLabels[constants.DataplaneModeLabel] = constants.DataplaneModeNone
+	}
+
+	// Default the network label for waypoints if not explicitly set in gateway's labels
+	network := d.injectConfig().Values.Struct().GetGlobal().GetNetwork()
+	if _, ok := gw.GetLabels()[label.TopologyNetwork.Name]; !ok && network != "" && isWaypointGateway {
+		input.InfrastructureLabels[label.TopologyNetwork.Name] = d.injectConfig().Values.Struct().GetGlobal().GetNetwork()
+	}
+}
+
+func extractInfrastructureMetadata(gwInfra *gatewayv1.GatewayInfrastructure, isLabel bool, gw gateway.Gateway) map[string]string {
```

Comment:
nit: don't love the bool arg; would prefer clearly named functions using a common base, but not a big deal

Reply:
:+1: done